### ISSUE TITLE
Upgrading from 1.6.0-alpha.3 to 1.6.0-beta.3 fails

### ISSUE THREAD

**Issue by brendandburns** [kubernetes/kubernetes#42976]
With 1.6.0-beta.3 I see a bunch of errors in the apiserver like:

```
unexpected ListAndWatch error: k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.go:215: Failed to list *api.Endpoints: couldn't get version/kind; json parse error: json: cannot unmarshal string into Go value of type struct { APIVersion string "json:\"apiVersion,omitempty\""; Kind string "json:\"kind,omitempty\"" }
```

My kubelets are all 1.5.2, it's possible that is causing the problem, but downgrading to 1.6.0-alpha.3 works correctly.

**Comment by ethernetdan** [comment]
@skriss @anhowe have you seen this failure before?

**Comment by liggitt** [comment]
that's coming from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/meta.go#L56

(prior history at https://github.com/kubernetes/kubernetes/blob/35a6bfbceed9d8bc779fbfe00963b7674ea68b77/pkg/runtime/serializer/json/meta.go)

I can get that error message if I pass a raw string (not valid json) to a json.Unmarshal call: https://play.golang.org/p/1-apy7ecWA

do you have your apiserver invocation args handy?

**Comment by smarterclayton** [comment]
Check your etcd storage (are you using etcd3, etcd2, json, protobuf) and whether the data for endpoints is consistent.  If you upgraded, then downgraded, you might get invalid data in etcd (unlikely, but one possible explanation).

**Comment by liggitt** [comment]
also, can you hit the endpoints API directly with kubectl or curl (both the unsecured and secured ports)

**Comment by rwehner** [comment]
I have what I think is a similar situation testing an upgrade from v1.5.2 to v1.6.0-beta.2 (leaving etcd unchanged at etcd2 and using `--storage-backend=etcd2`). Hopefully, it is the same issue and this context helps. Once the upgrade is complete and working, we attempt to downgrade the master node back to v1.5.2 (kubelets are still at v1.6.0-beta.2). The newly downgraded v1.5.2 kube-apiserver won't stay up. Begins with errors like:

```
I0314 15:15:44.459697       1 reflector.go:234] Listing and watching *api.ReplicationController from pkg/storage/cacher.go:201
E0314 15:15:44.463773       1 cacher.go:260] unexpected ListAndWatch error: pkg/storage/cacher.go:201: Failed to list *api.ConfigMap: couldn't get version/kind; json parse error: json: cannot unmarshal string into Go value of type struct { APIVersion string "json:\"apiVersion,omitempty\""; Kind string "json:\"kind,omitempty\"" }
```

The kube-apiserver finally dies (with `--v=8`) with this trace and fatal error: 
```
F0314 15:15:44.785634       1 controller.go:128] Unable to perform initial IP allocation check: unable to refresh the service IP block: couldn't get version/kind; json parse error: json: cannot unmarshal string into Go value of type struct { APIVersion string "json:\"apiVersion,omitempty\""; Kind string "json:\"kind,omitempty\"" }
E0314 15:15:44.803353       1 errors.go:63] apiserver received an error that is not an unversioned.Status: couldn't get
version/kind; json parse error: json: cannot unmarshal string into Go value of type struct { APIVersion string "json:\"a
piVersion,omitempty\""; Kind string "json:\"kind,omitempty\"" }
I0314 15:15:44.804046       1 panics.go:76] GET /api/v1/namespaces/kube-system/endpoints/kube-scheduler: (19.034086ms) 5
00
goroutine 502 [running]:
k8s.io/kubernetes/pkg/httplog.(*respLogger).recordStatus(0xc422793500, 0x1f4)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:219 +0xbb
k8s.io/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc422793500, 0x1f4)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:198 +0x35
k8s.io/kubernetes/pkg/apiserver/metrics.(*responseWriterDelegator).WriteHeader(0xc4227dc930, 0x1f4)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/metrics/metrics.go:117 +0x45
k8s.io/kubernetes/pkg/apiserver.writeNegotiated(0x5c59280, 0xc4200593e0, 0x0, 0x0, 0x343feed, 0x2, 0x5c51c40, 0xc4225272
88, 0xc4213650e0, 0x1f4, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go:463 +0x162
k8s.io/kubernetes/pkg/apiserver.errorNegotiated(0x5c2dc80, 0xc421f21300, 0x5c59280, 0xc4200593e0, 0x0, 0x0, 0x343feed, 0x2, 0x5c51c40, 0xc422527288, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go:480 +0x103
k8s.io/kubernetes/pkg/apiserver.(*RequestScope).err(0xc420841500, 0x5c2dc80, 0xc421f21300, 0x5c51c40, 0xc422527288, 0xc4213650e0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:88 +0xd7
k8s.io/kubernetes/pkg/apiserver.getResourceHandler.func1(0xc4227dc8a0, 0xc421e859e0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:113 +0x20d
k8s.io/kubernetes/pkg/apiserver/metrics.InstrumentRouteFunc.func1(0xc4227dc8a0, 0xc421e859e0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/metrics/metrics.go:101 +0x1f2
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc4208f4d80, 0x5c52380, 0xc422793500, 0xc4213650e0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:272 +0xba0
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).(k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.dispatch)-fm(0x5c52380, 0xc422793500, 0xc4213650e0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:120 +0x48
net/http.HandlerFunc.ServeHTTP(0xc420491590, 0x5c52380, 0xc422793500, 0xc4213650e0)
        /usr/local/go/src/net/http/server.go:1726 +0x44
net/http.(*ServeMux).ServeHTTP(0xc4209148d0, 0x5c52380, 0xc422793500, 0xc4213650e0)
        /usr/local/go/src/net/http/server.go:2022 +0x7f
k8s.io/kubernetes/pkg/genericapiserver/filters.WithPanicRecovery.func1(0x5c52380, 0xc422793500, 0xc4213650e0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/filters/panics.go:75 +0x24a
net/http.HandlerFunc.ServeHTTP(0xc4208f91d0, 0x7f9c6df43d20, 0xc422527278, 0xc4213650e0)
        /usr/local/go/src/net/http/server.go:1726 +0x44
k8s.io/kubernetes/pkg/apiserver/filters.WithRequestInfo.func1(0x7f9c6df43d20, 0xc422527278, 0xc4213650e0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/filters/requestinfo.go:45 +0x212
net/http.HandlerFunc.ServeHTTP(0xc4208f9260, 0x7f9c6df43d20, 0xc422527278, 0xc4213650e0)
        /usr/local/go/src/net/http/server.go:1726 +0x44
k8s.io/kubernetes/pkg/api.WithRequestContext.func1(0x7f9c6df43d20, 0xc422527278, 0xc4213650e0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/requestcontext.go:107 +0xef
net/http.HandlerFunc.ServeHTTP(0xc4208e0d60, 0x7f9c6df43d20, 0xc422527278, 0xc4213650e0)
        /usr/local/go/src/net/http/server.go:1726 +0x44
k8s.io/kubernetes/pkg/genericapiserver/filters.(*timeoutHandler).ServeHTTP.func1(0xc4208e0d80, 0x5c5b800, 0xc422527278, 0xc4213650e0, 0xc421e85860)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/filters/timeout.go:78 +0x8d
created by k8s.io/kubernetes/pkg/genericapiserver/filters.(*timeoutHandler).ServeHTTP
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/filters/timeout.go:80 +0x1db

logging error output: "k8s\x00\n\f\n\x02v1\x12\x06Status\x12\xdb\x01\n\x04\n\x00\x12\x00\x12\aFailure\x1a\xc4\x01couldn't get version/kind; json parse error: json: cannot unmarshal string into Go value of type struct { APIVersion string \"json:\\\"apiVersion,omitempty\\\"\"; Kind string \"json:\\\"kind,omitempty\\\"\" }\"\x000\xf4\x03\x1a\x00\"\x00"
 [[hyperkube/v1.5.2 (linux/amd64) kubernetes/08e0995/leader-election] 127.0.0.1:48524]
```

**Comment by liggitt** [comment]
@rwehner `k8s\x00` indicates you have protobuf data stored... what did you set for `--storage-media-type` when upgraded to 1.6? @smarterclayton, correct me if I'm wrong, but isn't protobuf in etcd2 a problem?

**Comment by rwehner** [comment]
@liggitt: Ah, that might explain things. I didn't set `--storage-media-type` at all when I upgraded; I just set ``--storage-backend`. I didn't realize both were needed. So my issue may not be related to Brendan's.

FWIW with the clue you gave (thanks!), I set the v1.5.2kube-apiserver to have both `--storage-backend=etcd2 --storage-media-type=application/vnd.kubernetes.protobuf` and it started working. Thanks! Sorry to have piggybacked what turns out to be an unrelated issue.

**Comment by ethernetdan** [comment]
@liggitt @mml do want to make changing defaults here a release blocker? otherwise, we should move to another milestone

**Comment by liggitt** [comment]
@mml @wojtek-t looks like a mix of `--storage-backend=etcd2` and `--storage-media-type=application/vnd.kubernetes.protobuf` is problematic. That results in base64-encoded protobuf in etcd2, which is difficult to migrate to straight protobuf in etcd3.

Because the defaults for both storage backend and storage media type changed in 1.6, I *really* want to prevent accidental mixing of etcd2 with protobuf storage serialization.

What do you think about preventing 1.6 from running in etcd2+protobuf mode? Did we support running 1.5 that way?

**Comment by smarterclayton** [comment]
I think 1.6 should prevent etcd2+protobuf, so users have to move to
something supportable.

**Comment by liggitt** [comment]
opened https://github.com/kubernetes/kubernetes/pull/43122 to guard against that combination

**Comment by wojtek-t** [comment]
Yes - etcd2 + protobuf is a bad combination (in both directions, i.e. it doesn't work when rollbacked from etcd3 + protobuf, as well as upgrade to etcd3 + protobuf doesn't work from it).

**Comment by liggitt** [comment]
@rwehner https://github.com/kubernetes/kubernetes/pull/43122 has merged which will prevent new protobuf data from being written into etcd2 (but allows existing data to be read).

@brendandburns do you have any more info about your env? want to make sure there's not another issue present

**Comment by liggitt** [comment]
@brendandburns also, is this reproducible upgrading from 1.5.x?

**Comment by liggitt** [comment]
closing this as fixed by https://github.com/kubernetes/kubernetes/pull/43122

@brendandburns if you can recreate with that fix or have more info that indicates a different issue, please reopen
